velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h

```diff
@@ -0,0 +1,342 @@
+                            .planNode();
+
+    core::PlanNodePtr plan;
+    if (minMaxByN_) {
+      plan =
+          PlanBuilder(planNodeIdGenerator, result->pool())
+              .localPartition({}, {expectedSource, actualSource})
+              .singleAggregation(
+                  groupingKeys_,
+                  {fmt::format("array_agg({}) as a", name_),
+                   "array_agg(expected) as e"})
+              .project({"remove_nulls(a) as a", "remove_nulls(e) as e"})
+              // Check that if actual result is not NULL, all elements in a[1]
+              // are also in e[1]. If the actual result is NULL, either the
+              // expected result is NULL too (when all rows in the group are
+              // masked out), or e[1] is empty (when all y in the group are
+              // NULL).
+              .project({fmt::format(
+                  "switch(cardinality(a) > 0, cardinality(e) > 0 and reduce(a[1], 0, (s, x) -> if (s < 0, -1, if (x is null, if (cardinality(e[1]) > cardinality(remove_nulls(e[1])), 1, -1), if (\"$internal$contains\"(e[1], x), 1, -1))), s -> s >= 0), cardinality(e) = 0 or cardinality(e[1]) = 0)",
+                  aggregateTypeSql_)})
+              .planNode();
+    } else {
+      plan =
+          PlanBuilder(planNodeIdGenerator, result->pool())
+              .localPartition({}, {expectedSource, actualSource})
+              .singleAggregation(
+                  groupingKeys_,
+                  {fmt::format("array_agg({}) as a", name_),
+                   "array_agg(expected) as e"})
+              .project({"remove_nulls(a) as a", "remove_nulls(e) as e"})
+              // Check that if actual result is not NULL, a[1]
+              // is contained in e[1]. If the actual result is NULL, either the
+              // expected result is NULL too (when all rows in the group are
+              // masked out), e[1] is empty (when all y in the group are
+              // NULL), or e[1] is [NULL] (when there is a NULL x associated
+              // with the min/max y).
+              .project(
+                  {"switch(cardinality(a) > 0, cardinality(e) > 0 and \"$internal$contains\"(e[1], a[1]), cardinality(e) = 0 or cardinality(e[1]) = 0 or cardinality(e[1]) > cardinality(remove_nulls(e[1])))"})
+              .planNode();
+    }
+
+    auto assertTrueResults = [&result](const core::PlanNodePtr& plan) {
+      auto contains = AssertQueryBuilder(plan).copyResults(result->pool());
+      auto numGroups = result->size();
+      VELOX_CHECK_EQ(numGroups, contains->size());
+
+      VectorPtr expected = BaseVector::createConstant(
+          BOOLEAN(), true, numGroups, result->pool());
+      velox::test::VectorMaker vectorMaker(result->pool());
+      auto expectedRow = vectorMaker.rowVector({expected});
+
+      return assertEqualResults({expectedRow}, {contains});
+    };
+    assertTrueResults(plan);
+
+    if (minMaxByN_) {
+      auto subsetSource =
```

Comment:
This part unfortunately was not clear to me (on why we are generating for subsetSource).

Reply:
subsetSource is just a PlanNode we're going to pass to localPartition together with actualSource. The goal of this localPartition is to combine the `subset_` and `result` together.

Reply:
Yes I get that, but why should result be in both subset and superset ?